### PR TITLE
Retry for adding remote cluster

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
**Please provide a description of this PR:**

Take over #39062

**To help us figure out who should review this PR, please put an X in all the areas that this PR affects.**

- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Installation
- [x] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

**Please check any characteristics that apply to this pull request.**

- [x] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

File: pilot/pkg/serviceregistry/kube/controller/controller.go [hunk 1]
```diff
@@ -151,7 +151,7 @@ func DetectEndpointMode(kubeClient kubelib.Client) EndpointMode {
 	useEndpointslice, ok := features.EnableEndpointSliceController()
 
 	// we have a client, and flag wasn't set explicitly, auto-detect
-	if kubeClient != nil && !ok && kubelib.IsAtLeastVersion(kubeClient, 21) {
+	if kubeClient != nil && !ok && !kubelib.IsLessThanVersion(kubeClient, 21) {
```
Comment (ramaraochavali) [inline pilot/pkg/serviceregistry/kube/controller/controller.go:154]: Why this change? Also can we reuse the same function/check at both places?

File: pkg/kube/multicluster/secretcontroller.go [hunk 2]
```diff
@@ -46,6 +48,7 @@ import (
 
 const (
 	MultiClusterSecretLabel = "istio/multiCluster"
+	maxRetries              = 15
```
Comment (ramaraochavali) [inline pkg/kube/multicluster/secretcontroller.go:51]: Is this some arbitrary number?

**Comment by hzxuzhonghu** [comment]
Something need to be fixed, this is all from https://github.com/istio/istio/pull/39062

File: pilot/pkg/serviceregistry/kube/controller/controller.go (hunk 1 above)
Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/controller.go:154]: This is because   when err happens , it may detect the k8s versions error, in this case if the k8s cluster is 1.20, then we could wrongly use endpointslice.
```
func IsAtLeastVersion(client Client, minorVersion uint) bool {
	clusterVersion, err := client.GetKubernetesVersion()
	if err != nil {
		return true
	}
	return IsKubeAtLeastOrLessThanVersion(clusterVersion, minorVersion, true)
}
```

**Comment by hzxuzhonghu** [comment]
@stevenctl Ptal

File: pilot/pkg/config/kube/crdclient/client.go [hunk 3]
```diff
@@ -423,24 +422,14 @@ func (cl *Client) kind(r config.GroupVersionKind) (*cacheHandler, bool) {
 // knownCRDs returns all CRDs present in the cluster, with timeout and retries.
 func knownCRDs(crdClient apiextensionsclient.Interface) (map[string]struct{}, error) {
 	var res *crd.CustomResourceDefinitionList
-	b := backoff.NewExponentialBackOff()
```
Comment (stevenctl) [inline pilot/pkg/config/kube/crdclient/client.go:426]: What's the reasoning behind this change?

File: pkg/kube/controllers/queue.go [hunk 4]
```diff
@@ -157,8 +157,9 @@ func (q Queue) processNextItem() bool {
 
 	err := q.workFn(key)
 	if err != nil {
-		if q.queue.NumRequeues(key) < q.maxAttempts {
-			q.log.Errorf("error handling %v, retrying: %v", key, err)
+		retryCount := q.queue.NumRequeues(key)
```
Comment (stevenctl) [inline pkg/kube/controllers/queue.go:160]: Does numRequeues get reset after we successfully process it? We might have > maxAttempts updates to the same key.

Comment (stevenctl) [reply pkg/kube/controllers/queue.go:160]: Looks like it will behave how we want.

**Comment by stevenctl** [comment]
LGTM, but why did we remove the knownCRDs retry? I don't think we go through the multi-cluster retries for all usages of that?

File: pilot/pkg/config/kube/crdclient/client.go (hunk 3 above)
Comment (hzxuzhonghu) [reply pilot/pkg/config/kube/crdclient/client.go:426]: We will  retry at the upper level

**Comment by hzxuzhonghu** [comment]
@stevenctl I confirmed， there are 3 kinds of usage:

1. during pilot bootstrap
2.  multicluster
3. CLI analysis
So I think removing the retry in function does not matter

**Comment by stevenctl** [comment]
Hmm. I think the pilot bootstrap case probably gets value but I don't have enough context about why we added the retry to begin with.

**Review by stevenctl — APPROVED** [review]
